src/Symfony/Component/DependencyInjection/Compiler/AutowirePass.php

```diff
@@ -246,20 +233,24 @@ private function autowireCalls(\ReflectionClass $reflectionClass, array $methodC
      *
      * @param \ReflectionMethod $reflectionMethod
      * @param array             $arguments
-     * @param int               $mode
      *
      * @return array The autowired arguments
      *
      * @throws RuntimeException
      */
-    private function autowireMethod(\ReflectionMethod $reflectionMethod, array $arguments, $mode)
+    private function autowireMethod(\ReflectionMethod $reflectionMethod, array $arguments)
     {
-        $didAutowire = false; // Whether any arguments have been autowired or not
+        $isConstructor = $reflectionMethod->isConstructor();
+
+        if (!$isConstructor && !$arguments && !$reflectionMethod->getNumberOfRequiredParameters()) {
+            throw new RuntimeException(sprintf('Cannot autowire service "%s": method %s::%s() has only optional arguments, thus must be wired explicitly.', $this->currentId, $reflectionMethod->class, $reflectionMethod->name));
```

Comment:
a method with no argument at all is considered as a getter, thus can't hit this method, see
https://github.com/symfony/symfony/pull/22030/files#diff-62df969ae028c559d33ffd256de1ac49R219